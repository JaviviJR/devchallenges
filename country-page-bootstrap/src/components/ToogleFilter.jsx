import PropTypes from 'prop-types';
import styles from './ToogleFilter.module.css';
import { useEffect, useState } from 'react';


function ToogleFilter({ label, filters, handler }) {
    const [selected, setSelected] = useState(filters.map(filter => filter.value));
    
    const isChecked = (value) => {
        return selected.includes(value);
    };

    const handleOnChange = (event) => {
        const { name } = event.target;
        const checked = isChecked(name);
        if (!checked) {
            setSelected([...selected, name]);
        } else {
            setSelected(selected.filter(item => item !== name));
        }
    };

    useEffect(() => {
        handler(selected);
    }, [selected]);

    return (
        <fieldset className={`d-flex flex-wrap`}>
            <legend className={`${styles.label1}`}>{ label }</legend>
            
            { filters.map(filter => (
                <div key={ filter.value }>
                    {/* <input 
                        type="checkbox"
                        className={`btn-check`}
                        name={ filter.value }
                        id={ filter.value }
                        // checked={true}
                        onChange={ handleOnChange }
                    />
                    <label 
                        // className={`btn ${styles.toogleLabel} ${selected.includes(filter.value) ? styles.active : ''}`}
                        className={`btn ${styles.toogleLabel}`}
                        htmlFor={ filter.value }
                    >
                        { filter.label }
                    </label> */}
                    <button 
                        type='button'
                        className={`btn ${styles.btnToogle} ${isChecked(filter.value) ? styles.active : ''}`}
                        name={ filter.value }
                        onClick={ handleOnChange }
                    >
                        { filter.label }
                    </button>
                </div>
            ))}

        </fieldset>
    );
}

export default ToogleFilter;

ToogleFilter.propTypes = {
    label: PropTypes.string.isRequired,
    filters: PropTypes.arrayOf(PropTypes.shape({
        value: PropTypes.string.isRequired,
        label: PropTypes.string.isRequired
    })).isRequired,
    handler: PropTypes.func.isRequired
};