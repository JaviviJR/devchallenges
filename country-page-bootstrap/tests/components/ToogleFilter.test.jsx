import { fireEvent, render, screen } from "@testing-library/react"
import ToogleFilter from "../../src/components/ToogleFilter"

describe('testing CheckFilter component', () => {
    const filtersTest = [
        { value: 'filter1Value', label: 'filter1Label' },
        { value: 'filter2Value', label: 'filter2Label' }
    ];

    const handlerMock = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should match with snapshot', () => {
        const { container } = render(
            <ToogleFilter
                label='labelTest'
                filters={ filtersTest }
                handler={ handlerMock }
            />
        );

        expect(container).toMatchSnapshot();
    })

    test('should show all filters', () => {
        render(
            <ToogleFilter
                label='labelTest'
                filters={ filtersTest }
                handler={ handlerMock }
            />
        );
        
        const filter1 = screen.getByRole('button', {name: 'filter1Label'});
        const filter2 = screen.getByRole('button', {name: 'filter2Label'});
        expect( filter1 ).toBeInTheDocument();
        expect( filter2 ).toBeInTheDocument();

        expect(filter1.name).toBe('filter1Value');
        expect(filter2.name).toBe('filter2Value');
    });

    test('should handle click', () => {
        render(
            <ToogleFilter
                label='labelTest'
                filters={ filtersTest }
                handler={ handlerMock }
            />
        );

        const filter1 = screen.getByRole('button', {name: 'filter1Label'});
        fireEvent.click(filter1);

        expect( filter1 ).toHaveClass('active')
        expect( handlerMock ).toHaveBeenLastCalledWith(['filter1Value']);
    });
})