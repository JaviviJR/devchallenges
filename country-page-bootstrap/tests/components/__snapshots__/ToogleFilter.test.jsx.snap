// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing CheckFilter component should match with snapshot 1`] = `
<div>
  <fieldset
    class="d-flex flex-wrap"
  >
    <legend
      class="label1"
    >
      labelTest
    </legend>
    <div>
      <button
        class="btn btnToogle "
        name="filter1Value"
        type="button"
      >
        filter1Label
      </button>
    </div>
    <div>
      <button
        class="btn btnToogle "
        name="filter2Value"
        type="button"
      >
        filter2Label
      </button>
    </div>
  </fieldset>
</div>
`;
